cardano-api/src/Cardano/Api/Governance/Poll.hs

```diff
@@ -0,0 +1,215 @@
+--
+
+data GovernancePollAnswer = GovernancePollAnswer
+    { govAnsPoll :: Hash GovernancePoll
+    , govAnsChoice :: Word
+    }
+  deriving Show
+
+instance HasTypeProxy GovernancePollAnswer where
+    data AsType GovernancePollAnswer = AsGovernancePollAnswer
+    proxyToAsType _ = AsGovernancePollAnswer
+
+instance SignableRepresentation GovernancePollAnswer where
+    getSignableRepresentation =
+      error "not implemented"
+
+instance SerialiseAsCBOR GovernancePollAnswer where
+    serialiseToCBOR =
+      error "not implemented"
+
+    deserialiseFromCBOR AsGovernancePollAnswer _bs =
+      error "not implemented"
+
+
+-- ----------------------------------------------------------------------------
+-- Governance Poll Witness
+--
+
+data GovernancePollWitness
+    = GovernancePollWitnessVRF
+        (VerKeyVRF StandardCrypto)
+        (VRF.CertVRF (VRF StandardCrypto))
+    | GovernancePollWitnessColdKey
+        (VKey 'Witness StandardCrypto)
+        (SignedDSIGN StandardCrypto GovernancePollAnswer)
+  deriving Show
+
+instance HasTypeProxy GovernancePollWitness where
+    data AsType GovernancePollWitness = AsGovernancePollWitness
+    proxyToAsType _ = AsGovernancePollWitness
+
+instance SerialiseAsCBOR GovernancePollWitness where
+    serialiseToCBOR =
+      error "not implemented"
+
+    deserialiseFromCBOR AsGovernancePollWitness _bs =
+      error "not implemented"
+
+signPollAnswerWith
+  :: GovernancePollAnswer
+  -> Either (SignKeyVRF StandardCrypto) (SignKeyDSIGN StandardCrypto)
+  -> GovernancePollWitness
+signPollAnswerWith answer =
+  either
+    (\sk -> GovernancePollWitnessVRF
+      (VRF.deriveVerKeyVRF sk)
+      (snd $ VRF.evalVRF () answer sk)
+    )
+    (\sk -> GovernancePollWitnessColdKey
+      (VKey (DSIGN.deriveVerKeyDSIGN sk))
+      (signedDSIGN @StandardCrypto sk answer)
+    )
+
+-- ----------------------------------------------------------------------------
+-- Governance Poll Verification
+--
+
+data GovernancePollError
+  = ErrGovernancePollMismatch
+  | ErrGovernancePollInvalidAnswer
+  | ErrGovernancePollInvalidWitness
+  deriving Show
+
+renderGovernancePollError :: GovernancePollError -> Text
+renderGovernancePollError err =
+  case err of
+    ErrGovernancePollMismatch ->
+      "Answer's poll doesn't match provided poll (hash mismatch)."
+    ErrGovernancePollInvalidAnswer ->
+      "Invalid answer: not part of the poll."
+    ErrGovernancePollInvalidWitness ->
+      "Invalid witness for the answer: the proof / signature doesn't hold."
+
+verifyPollAnswer
+  :: GovernancePoll
+  -> GovernancePollAnswer
+  -> GovernancePollWitness
+  -> Either GovernancePollError ()
+verifyPollAnswer poll answer witness = do
+  when (hashGovernancePoll poll /= govAnsPoll answer) $
+    Left ErrGovernancePollMismatch
+
+  when (govAnsChoice answer >= fromIntegral (length (govPollAnswers poll))) $
+    Left ErrGovernancePollInvalidAnswer
+
+  unless isValid $
+    Left ErrGovernancePollInvalidWitness
+
+  pure ()
```

Comment:
There is a `hlint` warning here as well.

Reply:
Done.